From 3b7424f3327304988431eca44adf92e59999aa2b Mon Sep 17 00:00:00 2001
From: David Wu <david.wu@rock-chips.com>
Date: Tue, 5 Sep 2017 18:16:13 +0800
Subject: [PATCH] pinctrl: rockchip: Make the gpio base started from 1000

If the gpio base is started from 1000, the "gpio = 0" will
be invalid, that can avoid something unforeseen. The real
pin number is "gpio number - 1000".

If you cat the gpio log, you will see the log like this:
rk3399:/ # cat d/gpio
GPIOs 1000-1031, platform/pinctrl, gpio0:
 gpio-1004 (                    |bt_default_wake_host) in  lo
 gpio-1005 (                    |power               ) in  hi
 gpio-1009 (                    |bt_default_reset    ) out lo
 gpio-1010 (                    |reset               ) out lo
 gpio-1011 (                    |?                   ) out hi

GPIOs 1032-1063, platform/pinctrl, gpio1:
 gpio-1034 (                    |int-n               ) in  hi
 gpio-1035 (                    |vbus-5v             ) out lo
 gpio-1036 (                    |vbus-5v             ) out lo
 gpio-1045 (                    |enable              ) out hi
 gpio-1046 (                    |vsel                ) out lo
 gpio-1049 (                    |vsel                ) out lo
 gpio-1056 (                    |int-n               ) in  hi

GPIOs 1064-1095, platform/pinctrl, gpio2:
 gpio-1083 (                    |bt_default_rts      ) in  hi
 gpio-1090 (                    |bt_default_wake     ) in  hi

GPIOs 1096-1127, platform/pinctrl, gpio3:
 gpio-1111 (                    |mdio-reset          ) out hi

GPIOs 1128-1159, platform/pinctrl, gpio4:
 gpio-1150 (                    |?                   ) out hi
 gpio-1153 (                    |vcc5v0_host         ) out hi
 gpio-1156 (                    |hp det              ) in  hi

Change-Id: I744ddc1df6075b0a044d65c65622e2a59f3a332e
Signed-off-by: David Wu <david.wu@rock-chips.com>
---
 drivers/pinctrl/pinctrl-rockchip.c | 6 +++---
 include/asm-generic/gpio.h         | 4 +++-
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/drivers/pinctrl/pinctrl-rockchip.c b/drivers/pinctrl/pinctrl-rockchip.c
index 127812ffc96f..289305ad414f 100644
--- a/drivers/pinctrl/pinctrl-rockchip.c
+++ b/drivers/pinctrl/pinctrl-rockchip.c
@@ -2208,11 +2208,11 @@ static int rockchip_pmx_gpio_set_direction(struct pinctrl_dev *pctldev,
 	int pin;
 
 	chip = range->gc;
-	pin = offset - chip->base;
+	pin = offset - range->pin_base;
 	dev_dbg(info->dev, "gpio_direction for pin %u as %s-%d to %s\n",
 		 offset, range->name, pin, input ? "input" : "output");
 
-	return _rockchip_pmx_gpio_set_direction(chip, offset - chip->base,
+	return _rockchip_pmx_gpio_set_direction(chip, pin,
 						input);
 }
 
@@ -3014,7 +3014,7 @@ static int rockchip_gpiolib_register(struct platform_device *pdev,
 		bank->gpio_chip = rockchip_gpiolib_chip;
 
 		gc = &bank->gpio_chip;
-		gc->base = bank->pin_base;
+		gc->base = ARCH_GPIO_BASE + bank->pin_base;
 		gc->ngpio = bank->nr_pins;
 		gc->dev = &pdev->dev;
 		gc->of_node = bank->of_node;
diff --git a/include/asm-generic/gpio.h b/include/asm-generic/gpio.h
index 40ec1433f05d..56bfbb4e497f 100644
--- a/include/asm-generic/gpio.h
+++ b/include/asm-generic/gpio.h
@@ -25,6 +25,8 @@
  * actually an estimate of a board-specific value.
  */
 
+#define ARCH_GPIO_BASE		1000
+
 #ifndef ARCH_NR_GPIOS
 #define ARCH_NR_GPIOS		512
 #endif
@@ -40,7 +42,7 @@
 
 static inline bool gpio_is_valid(int number)
 {
-	return number >= 0 && number < ARCH_NR_GPIOS;
+	return number >= ARCH_GPIO_BASE && number < ARCH_NR_GPIOS + ARCH_GPIO_BASE;
 }
 
 struct device;
-- 
2.35.3

